src/vkex/Device.h

```diff
@@ -228,6 +249,8 @@ class CPhysicalDevice {
     VkPhysicalDeviceSampleLocationsPropertiesEXT sample_locations_properties;
   } m_extension_owned_properties;
 
+  PhysicalDeviceExtensionFeatures m_extension_owned_features;
```

Comment:
I'm not sure if I'm missing something, or noticing something you overlooked, so perhaps this question will help clarify things for one of us: what is the reason for this m_extension_owned_features AND the extension_features in DeviceCreateInfo? I'd have assumed one copy of this stuff would be enough and can't quite wrap my ahead around why both are needed (eg. what the difference between them is).

Reply:
You're not missing anything, Vulkan is just fucking awful to use in this regard.

`m_extension_owned_features` belongs to the PhysicalDevice. These are features that are _possible_ on an implementation and tied to device extensions. But since you need to know this info BEFORE you create the device, you get to query them from the physical device via pNext. So here, I'm just querying all the capabilities independent of the extensions that are actually enabled. If the extension that owns some of these features doesn't exist, then the feature should come back with everything flipped off. I thought about filtering them via available extensions, but I figured I could defer that check to _after_ the requested extension list is filtered to what is available.

Sidebar: Of course, when I actually ran this code, I found out that my Intel laptop returns valid bits for shader_float16_int8, but the extension isn't exposed, wtf?

The DeviceCreateInfo copy of PhysicalDeviceExtensionFeatures is the 'filtered' version based on what the implementation exposes and the extensions requested.

I did think about making the float16/int8 stuff a default request for vkex, but I figured I should defer that conversation to whenever I make vkex a submodule, and talk to Hai about how he wants it set up.

Thanks for coming to my TED talk

Reply:
😂

got it, OK then ya I think that makes just enough sense.